imap/pushsub_db.h

```diff
@@ -0,0 +1,120 @@
+/* pushsub_db.h -- abstract interface for per-user PushSubcription database
+ *
+ * Copyright (c) 1994-2022 Carnegie Mellon University.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *
+ * 3. The name "Carnegie Mellon University" must not be used to
+ *    endorse or promote products derived from this software without
+ *    prior written permission. For permission or any legal
+ *    details, please contact
+ *      Carnegie Mellon University
+ *      Center for Technology Transfer and Enterprise Creation
+ *      4615 Forbes Avenue
+ *      Suite 302
+ *      Pittsburgh, PA  15213
+ *      (412) 268-7393, fax: (412) 268-7395
+ *      innovation@andrew.cmu.edu
+ *
+ * 4. Redistributions of any form whatsoever must retain the following
+ *    acknowledgment:
+ *    "This product includes software developed by Computing Services
+ *     at Carnegie Mellon University (http://www.cmu.edu/computing/)."
+ *
+ * CARNEGIE MELLON UNIVERSITY DISCLAIMS ALL WARRANTIES WITH REGARD TO
+ * THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
+ * AND FITNESS, IN NO EVENT SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE
+ * FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
+ * AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
+ * OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#ifndef PUSHSUB_DB_H
+#define PUSHSUB_DB_H
+
+#include <config.h>
+
+#include "mailbox.h"
+#include "sqldb.h"
+
+struct pushsub_db;
+
+#define PUSHSUB_CREATE 0x01
+#define PUSHSUB_TRUNC  0x02
+
+struct pushsub_data {
+    unsigned rowid;
+    const char *mailbox;
+    uint32_t imap_uid;
+    const char *id;
+    const char *subscription;
+    time_t expires;
+    unsigned isverified;
+    int alive;
+};
+
```

Comment:
Reordaring the elements to
```c
struct pushsub_data_new {
    const char *mailbox;
    const char *id;
    const char *subscription;
    unsigned rowid;
    unsigned isverified;
    time_t expires;
    int alive;
};
```
will reduce the data needed for the structure. (but it depends how the struct is used)

```c
#include <time.h>
#include <stdio.h>
#include <stdint.h>

struct pushsub_data {
    unsigned rowid;
    const char *mailbox;
    uint32_t imap_uid;
    const char *id;
    const char *subscription;
    time_t expires;
    unsigned isverified;
    int alive;
};

struct pushsub_data_new {
    const char *mailbox;
    const char *id;
    const char *subscription;
    unsigned rowid;
    unsigned isverified;
    time_t expires;
    int alive;
};


int main() {
  printf("%lu\n", sizeof(struct pushsub_data));
  printf("%lu\n", sizeof(struct pushsub_data_new));
  return 0;
}
```
prints
```
$ clang -Wpadded pad.c -o pad && ./pad
pad.c:7:17: warning: padding struct 'struct pushsub_data' with 4 bytes to align 'mailbox' [-Wpadded]
    const char *mailbox;
                ^
pad.c:9:17: warning: padding struct 'struct pushsub_data' with 4 bytes to align 'id' [-Wpadded]
    const char *id;
                ^
pad.c:16:8: warning: padding size of 'struct pushsub_data_new' with 4 bytes to alignment boundary [-Wpadded]
struct pushsub_data_new {
       ^
3 warnings generated.
56
48
```

Reply:
The existing struct matches those in  the other *_db.h files.  I'm not sure we are too worried about 8 bytes